system/containers.toit

```diff
@@ -118,18 +118,98 @@ class ContainerImageFlash extends ContainerImage:
+        --minor=ContainersService.MINOR
+    install
+
+  handle client/int index/int arguments/any -> any:
+    if index == ContainersService.LIST_IMAGES_INDEX:
+      return list_images
+    if index == ContainersService.START_IMAGE_INDEX:
+      return start_image (uuid.Uuid arguments)
+    if index == ContainersService.UNINSTALL_IMAGE_INDEX:
+      return uninstall_image (uuid.Uuid arguments)
+    if index == ContainersService.IMAGE_WRITER_OPEN_INDEX:
+      return image_writer_open client arguments
+    if index == ContainersService.IMAGE_WRITER_WRITE_INDEX:
+      return image_writer_write client arguments[0] arguments[1]
+    if index == ContainersService.IMAGE_WRITER_COMMIT_INDEX:
+      return (image_writer_commit client arguments).to_byte_array
+    if index == ContainersService.IMAGE_WRITER_CLOSE_INDEX:
+      return image_writer_close client arguments
+    unreachable
+
+  abstract image_registry -> FlashRegistry
+  abstract images -> List
+  abstract add_flash_image allocation/FlashAllocation -> ContainerImage
+  abstract lookup_image id/uuid.Uuid -> ContainerImage?
+
+  abstract register_descriptor client/int descriptor/Object -> int
+  abstract lookup_descriptor client/int handle/int -> any
+  abstract unregister_descriptor client/int handle/int -> none
+
+  list_images -> List:
+    return images.map --in_place: | image/ContainerImage |
+      image.id.to_byte_array
+
+  start_image id/uuid.Uuid -> int?:
+    image/ContainerImage? := lookup_image id
+    if not image: return null
+    return image.start.id
+
+  uninstall_image id/uuid.Uuid -> none:
+    image/ContainerImage? := lookup_image id
+    if not image: return
+    image.delete
+
+  image_writer_open size/int -> int:
+    unreachable  // <-- TODO(kasper): Nasty.
+
+  image_writer_write handle/int bytes/ByteArray -> none:
+    unreachable  // <-- TODO(kasper): Nasty.
+
+  image_writer_commit handle/int -> uuid.Uuid:
+    unreachable  // <-- TODO(kasper): Nasty.
+
+  image_writer_close handle/int -> none:
+    unreachable  // <-- TODO(kasper): Nasty.
+
+  image_writer_open client/int size/int -> int?:
+    relocated_size := size - (size / IMAGE_CHUNK_SIZE) * IMAGE_WORD_SIZE
+    reservation := image_registry.reserve relocated_size
+    if reservation == null: throw "FIXME: kasper"
+    writer := ContainerImageWriter reservation
+    return register_descriptor client writer
+
+  image_writer_write client/int handle/int bytes/ByteArray -> none:
+    writer := lookup_descriptor client handle
+    writer.write bytes
+
+  image_writer_commit client/int handle/int -> uuid.Uuid:
+    writer := lookup_descriptor client handle
+    image := add_flash_image writer.commit
+    return image.id
+
+  image_writer_close client/int handle/int -> none:
+    writer := lookup_descriptor client handle
+    unregister_descriptor client handle
+    writer.close
+
+class ContainerManager extends ContainersServiceDefinition implements SystemMessageHandler_:
```

Comment:
This class and ContainersServiceDefinition would benefit from a line or two explaining what they are for.

Reply:
You're right. I'm still really messing with things right now, but they do deserve toitdoc comments.